qiskit_nature/problems/second_quantization/molecular/integrals_calculators/angular_momentum_integrals_calculator.py

```diff
@@ -0,0 +1,129 @@
+
+import itertools
+from typing import Tuple, List
+
+import numpy as np
+
+
+def calc_total_ang_momentum_ints(num_modes: int) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Calculates 1- and 2-body integrals for a total angular momentum.
+
+    Args:
+        num_modes (int): Number of modes.
+
+    Returns:
+        Tuple(numpy.ndarray, numpy.ndarray): Tuple of 1- and 2-body integrals for a total angular
+        momentum.
+    """
+    x_h1, x_h2 = _calc_s_x_squared_ints(num_modes)
+    y_h1, y_h2 = _calc_s_y_squared_ints(num_modes)
+    z_h1, z_h2 = _calc_s_z_squared_ints(num_modes)
+    h_1 = x_h1 + y_h1 + z_h1
+    h_2 = x_h2 + y_h2 + z_h2
+
+    return h_1, h_2
+
+
+def _calc_s_x_squared_ints(num_modes: int) -> Tuple[np.ndarray, np.ndarray]:
+    return _calc_squared_ints(num_modes, _modify_s_x_squared_ints_neq, _modify_s_x_squared_ints_eq)
+
+
+def _calc_s_y_squared_ints(num_modes: int) -> Tuple[np.ndarray, np.ndarray]:
+    return _calc_squared_ints(num_modes, _modify_s_y_squared_ints_neq, _modify_s_y_squared_ints_eq)
+
+
+def _calc_s_z_squared_ints(num_modes: int) -> Tuple[np.ndarray, np.ndarray]:
+    return _calc_squared_ints(num_modes, _modify_s_z_squared_ints_neq, _modify_s_z_squared_ints_eq)
+
+
+def _calc_squared_ints(num_modes: int, func_neq, func_eq) -> Tuple[np.ndarray, np.ndarray]:
+    num_modes_2 = num_modes // 2
+    h_1 = np.zeros((num_modes, num_modes))
+    h_2 = np.zeros((num_modes, num_modes, num_modes, num_modes))
+
+    for p, q in itertools.product(range(num_modes_2), repeat=2):  # pylint: disable=invalid-name
+        if p != q:
+            h_2 = func_neq(h_2, p, q, num_modes_2)
+        else:
+            h_2 = func_eq(h_2, p, num_modes_2)
+            h_1[p, p] += 1.0
+            h_1[p + num_modes_2, p + num_modes_2] += 1.0
+    h_1 *= 0.25
+    h_2 *= 0.25
+    return h_1, h_2
+
+
+def _modify_s_x_squared_ints_neq(h_2: np.ndarray, p_ind: int, q_ind: int,
+                                 num_modes_2: int) -> np.ndarray:
+    indices = [(p_ind, p_ind + num_modes_2, q_ind, q_ind + num_modes_2),
+               (p_ind + num_modes_2, p_ind, q_ind, q_ind + num_modes_2),
+               (p_ind, p_ind + num_modes_2, q_ind + num_modes_2, q_ind),
+               (p_ind + num_modes_2, p_ind, q_ind + num_modes_2, q_ind)]
+    values = [1, 1, 1, 1]
+    return _modify_s_squared_ints(h_2, indices, values)
+
+
+def _modify_s_x_squared_ints_eq(h_2: np.ndarray, p_ind: int, num_modes_2: int) -> np.ndarray:
+    indices = [(p_ind, p_ind + num_modes_2, p_ind, p_ind + num_modes_2),
+               (p_ind + num_modes_2, p_ind, p_ind + num_modes_2, p_ind),
+               (p_ind, p_ind, p_ind + num_modes_2, p_ind + num_modes_2),
+               (p_ind + num_modes_2, p_ind + num_modes_2, p_ind, p_ind)]
+    values = [-1, -1, -1, -1]
+    return _modify_s_squared_ints(h_2, indices, values)
+
+
+def _modify_s_y_squared_ints_neq(h_2: np.ndarray, p_ind: int, q_ind: int,
```

Comment:
It might be helpful to have comments in there here to help the reader as 'modify' is such a generic term. For my part, even though you might have said there was duplication in the code, it seems to me that the original was easier to follow/read - but maybe that's partly because I am more used to seeing it.

Reply:
I have to agree :+1: @dlasecki if you could add a few comments (or even proper docstrings) to these methods that would be great :slightly_smiling_face:

Reply:
Sure. Hopefully the situation is improved now.

Reply:
I think this is resolved with the comments added.